### ISSUE TITLE
Running with DirectML gives "RuntimeError: The parameter is incorrect."

### ISSUE THREAD

**Issue by Acly** [cubiq/ComfyUI_IPAdapter_plus#10]
I'm running ComfyUI with `--directml` and using the example workflow from the readme.

There is some issue with batch size, I get "RuntimeError: The parameter is incorrect."  [here](https://github.com/cubiq/ComfyUI_IPAdapter_plus/blob/main/IPAdapterPlus.py#L188). Stack trace below.
It looks like `b = n.shape[0] = 1`, so 0 is passed to `repeat(b//2, ...)`. I guess the expected value would be `b = 2` but not really sure why.

Setting batch size to 2 will no longer trigger the error, produces 2 images, but only the second one is affected by IP adapter. Also when passing b to repeat instead of b//2. Any ideas?

Additional notes:
* `torch.autocast("cuda")` is maybe also a problem, but in my case everything is float32
* everything works fine with CUDA or CPU

Stack:
```
Traceback (most recent call last):
  File "C:\Dev\DML_Comfy\ComfyUI\execution.py", line 152, in recursive_execute
    output_data, output_ui = get_output_data(obj, input_data_all)
  File "C:\Dev\DML_Comfy\ComfyUI\execution.py", line 82, in get_output_data
    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
  File "C:\Dev\DML_Comfy\ComfyUI\execution.py", line 75, in map_node_over_list
    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
  File "C:\Dev\DML_Comfy\ComfyUI\nodes.py", line 1236, in sample
    return common_ksampler(model, seed, steps, cfg, sampler_name, scheduler, positive, negative, latent_image, denoise=denoise)
  File "C:\Dev\DML_Comfy\ComfyUI\nodes.py", line 1206, in common_ksampler
    samples = comfy.sample.sample(model, noise, steps, cfg, sampler_name, scheduler, positive, negative, latent_image,
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\sample.py", line 93, in sample
    samples = sampler.sample(noise, positive_copy, negative_copy, cfg=cfg, latent_image=latent_image, start_step=start_step, last_step=last_step, force_full_denoise=force_full_denoise, denoise_mask=noise_mask, sigmas=sigmas, callback=callback, disable_pbar=disable_pbar, seed=seed)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\samplers.py", line 742, in sample
    samples = getattr(k_diffusion_sampling, "sample_{}".format(self.sampler))(self.model_k, noise, sigmas, extra_args=extra_args, callback=k_callback, disable=disable_pbar)
  File "C:\Dev\DML_Comfy\python\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\k_diffusion\sampling.py", line 580, in sample_dpmpp_2m
    denoised = model(x, sigmas[i] * s_in, **extra_args)
  File "C:\Dev\DML_Comfy\python\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\samplers.py", line 323, in forward
    out = self.inner_model(x, sigma, cond=cond, uncond=uncond, cond_scale=cond_scale, cond_concat=cond_concat, model_options=model_options, seed=seed)
  File "C:\Dev\DML_Comfy\python\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\k_diffusion\external.py", line 125, in forward
    eps = self.get_eps(input * c_in, self.sigma_to_t(sigma), **kwargs)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\k_diffusion\external.py", line 151, in get_eps
    return self.inner_model.apply_model(*args, **kwargs)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\samplers.py", line 311, in apply_model
    out = sampling_function(self.inner_model.apply_model, x, timestep, uncond, cond, cond_scale, cond_concat, model_options=model_options, seed=seed)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\samplers.py", line 289, in sampling_function
    cond, uncond = calc_cond_uncond_batch(model_function, cond, uncond, x, timestep, max_total_area, cond_concat, model_options)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\samplers.py", line 265, in calc_cond_uncond_batch
    output = model_function(input_x, timestep_, **c).chunk(batch_chunks)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\model_base.py", line 63, in apply_model
    return self.diffusion_model(xc, t, context=context, y=c_adm, control=control, transformer_options=transformer_options).float()
  File "C:\Dev\DML_Comfy\python\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\ldm\modules\diffusionmodules\openaimodel.py", line 626, in forward
    h = forward_timestep_embed(module, h, emb, context, transformer_options)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\ldm\modules\diffusionmodules\openaimodel.py", line 56, in forward_timestep_embed
    x = layer(x, context, transformer_options)
  File "C:\Dev\DML_Comfy\python\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\ldm\modules\attention.py", line 693, in forward
    x = block(x, context=context[i], transformer_options=transformer_options)
  File "C:\Dev\DML_Comfy\python\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\ldm\modules\attention.py", line 525, in forward
    return checkpoint(self._forward, (x, context, transformer_options), self.parameters(), self.checkpoint)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\ldm\modules\diffusionmodules\util.py", line 123, in checkpoint
    return func(*inputs)
  File "C:\Dev\DML_Comfy\ComfyUI\comfy\ldm\modules\attention.py", line 622, in _forward
    n = attn2_replace_patch[block_attn2](n, context_attn2, value_attn2, extra_options)
  File "C:\Dev\DML_Comfy\ComfyUI\custom_nodes\ComfyUI_IPAdapter_plus\IPAdapterPlus.py", line 188, in __call__
    uncond_cond = torch.cat([uncond.repeat(b//2, 1, 1), cond.repeat(b//2, 1, 1)], dim=0)
RuntimeError: The parameter is incorrect.
```

**Comment by cubiq** [comment]
what version of directml? have you tried updating (possibly to a latest dev release)?

**Comment by Acly** [comment]
Python 3.10.11
torch 2.0.0
torch-directml 0.2.0.dev230426

This is the latest version on pypi, I can't find anything newer. It was built with torch 2.0.0 so doesn't work for 2.0.1

**Comment by cubiq** [comment]
I have an idea what could cause this but I cannot check on directml. I'm working on a refactor of the attention code so hopefully that will fix this issue as well

**Comment by cubiq** [comment]
can you try the latest version?

**Comment by Acly** [comment]
Thanks for looking into this! I can confirm the batch size fix is working for DirectML. I tried various batch sizes and all images are consistent now.

I had to revert the autocast to "cuda" though, it's not working as it is:
* With "cpu" I get "RuntimeError: Currently, AutocastCPU only support Bfloat16 as the autocast_cpu_dtype"
  This is an exception so it pops up as an error and stops executing
* With "cuda" I get "UserWarning: User provided device_type of 'cuda', but CUDA is not available. Disabling"
  This is just a warning, and it runs fine anyway.

I don't know if the "cpu" mode is useful at all (even when running on cpu). I don't think Comfy uses fp16 for either CPU or DirectML at the moment, so it's probably fine if it's ignored  for now.
